opal/corelib/array.rb

```diff
@@ -1357,16 +1368,27 @@ def keep_if(&block)
     return enum_for(:keep_if) { size } unless block_given?
 
     %x{
-      for (var i = 0, length = self.length, value; i < length; i++) {
-        value = block(self[i]);
+      var value, raised = null, updated = new Array(self.length);
 
-        if (value === false || value === nil) {
-          self.splice(i, 1);
+      for (var i = 0, i2 = 0, length = self.length; i < length; i++) {
+        try {
+          value = Opal.yield1(block, self[i])
+        } catch(error) {
+          raised = error;
+        }
 
-          length--;
-          i--;
+        if (raised || Opal.truthy(value)) {
+          updated[i2] = self[i]
+          i2 += 1;
         }
       }
+
+      if (i2 !== i) {
+        self.splice.apply(self, [0, updated.length].concat(updated));
+        self.splice(i2, updated.length);
+      }
+
+      if (raised) throw raised;
```

Comment:
MRI aborts iteration once the error has been thrown:

```ruby
[1,2,3].keep_if { |i| p i; raise 'a' }
1
# ...backtrace
RuntimeError (a)
```

So I guess this check-and-throw should be somewhere above (same for other methods)